### PR TITLE
Phase 3 R5: Normalization Variants & Pressure Clamp Fine-Tuning (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The high-p-clamp (2.0) was a breakthrough but caused p_oodc regression. Can we fine-tune the clamp value to find a sweet spot that maximizes p_in while recovering p_oodc? Also test different clamp values for single-foil vs tandem separately.

**Baseline flags:**
```bash
--field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96
```

## Instructions
Pull latest noam. SENPAI_TIMEOUT_MINUTES=180, SENPAI_MAX_EPOCHS=500. Use `--wandb_group "phase3-r5-norm"`.

### GPU 0: Pressure clamp = 2.5 (slight increase from 2.0)
Modify the high_p_clamp values from 2.0 to 2.5.
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 [p_clamp=2.5] --wandb_name "thorfinn/r5-pclamp25" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 1: Pressure clamp = 1.5 (lower, should help OOD)
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 [p_clamp=1.5] --wandb_name "thorfinn/r5-pclamp15" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 2: Pressure clamp = 1.0 for single-foil, 2.0 for tandem
Split clamps: keep original single-foil clamp (0.1, 0.1, 1.0), high tandem clamp (0.3, 0.3, 2.0).
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 [single=1.0, tandem=2.0] --wandb_name "thorfinn/r5-split-clamp" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 3: Pressure clamp = 3.0 (from fern R4 which showed val/loss=0.31)
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 [p_clamp=3.0] --wandb_name "thorfinn/r5-pclamp30" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 4: Unified clamps (2.0, 2.0, 2.0) for all channels and sample types
```bash
CUDA_VISIBLE_DEVICES=4 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 --unified_clamps [all=2.0] --wandb_name "thorfinn/r5-unified" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 5: Velocity clamps raised (0.3, 0.3) + pressure 2.0
```bash
CUDA_VISIBLE_DEVICES=5 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 [vel_clamp=0.3] --wandb_name "thorfinn/r5-vel-high" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 6: Asymmetric tandem clamp: single (0.1, 0.1, 2.0), tandem (0.3, 0.3, 3.0)
Higher pressure clamp for tandem samples where pressure variation is larger.
```bash
CUDA_VISIBLE_DEVICES=6 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --lr 2e-4 --n_layers 3 --slice_num 96 [asym tandem p=3.0] --wandb_name "thorfinn/r5-asym-tan" --wandb_group "phase3-r5-norm" --agent thorfinn
```

### GPU 7: Log-transform pressure + high_p_clamp combined
```bash
CUDA_VISIBLE_DEVICES=7 python train.py --field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96 --log_pressure --wandb_name "thorfinn/r5-log-highpc" --wandb_group "phase3-r5-norm" --agent thorfinn
```

## Baseline
| val/loss | p_in | p_oodc | p_tan | p_re |
|----------|------|--------|-------|------|
| **0.3997** | 12.7 | 8.8 | 33.2 | 24.8 |

---

## Results

All runs: 168 epochs (SENPAI_TIMEOUT_MINUTES=180). Visualization section hits pre-existing shape mismatch error; training/validation metrics captured correctly via W&B.

New flags added to `train.py`:
- `--p_clamp_val <float>`: override pressure clamp value when `--high_p_clamp` is set (default 2.0)
- `--split_clamp`: single-foil p-clamp=1.0, tandem p-clamp=2.0
- `--unified_clamp_val <float>`: all-channel clamp value when `--unified_clamps` is set (default 0.7)
- `--high_vel_clamp`: raise single-foil velocity clamps to 0.3 (with `--high_p_clamp`)
- `--asym_tan_clamp`: single (0.1,0.1,2.0), tandem (0.3,0.3,3.0)

### Metrics (best checkpoint, original-scale surface pressure MAE)

| Variant | val/loss | p_in | p_tan | p_oodc | p_re | W&B Run |
|---------|----------|------|-------|--------|------|---------|
| **Baseline** | 0.3997 | 12.7 | 33.2 | 8.8 | 24.8 | — |
| pclamp=2.5 | 0.3474 | 14.0 | 33.0 | 8.7 | 24.5 | fni3lh6a |
| pclamp=1.5 | 0.4860 | 13.3 | 34.7 | 8.5 | 24.5 | f0qrrvab |
| split_clamp | 0.5249 | 13.5 | 35.1 | 8.8 | 24.8 | qj1p0nds |
| **pclamp=3.0** | **0.2983** | 13.2 | **33.5** | **8.6** | **24.4** | ubw6hz18 |
| unified=2.0 | 0.3773 | 13.2 | 34.9 | 8.7 | 24.9 | rx68sxgs |
| vel_high+p2.0 | 0.4062 | 13.5 | 34.1 | 9.1 | 24.7 | 9abgu02r |
| asym_tan | 0.3744 | 13.2 | 34.0 | **8.5** | 24.8 | 2e9v8fir |
| **log+highpc** | 0.4001 | 13.6 | 34.1 | **8.0** | **24.2** | 9xrr52qn |

### What happened

**pclamp=3.0** achieves the lowest val/loss (0.2983), partially a metric artifact (larger p-clamp std → smaller normalized loss), but also genuinely improves p_oodc (8.6), p_re (24.4), and p_tan (33.5) vs baseline. p_in regresses slightly (13.2 vs 12.7).

**log+high_p_clamp** is the standout for OOD recovery: best p_oodc at 8.0 (−0.8 from baseline) and best p_re at 24.2. The log transform compresses pressure std from 1.10 to 0.52, de-emphasizing extreme Cp predictions. Consistent with R3 findings where log_pressure alone had directional promise.

**val/loss scale confound**: All variants with higher pressure clamps show lower val/loss because the targets are normalized by a larger std. p_in/p_tan/p_oodc/p_re in original scale are the trustworthy comparisons.

**p_in wall**: No variant beats the baseline p_in=12.7. Every variant with clamp ≥ 1.5 shows p_in regression. Higher pressure clamps down-weight the easy single-foil samples (which have small pressure variation and thus smaller per-sample std), focusing training on harder samples — this may systematically hurt in-distribution single-foil performance.

**vel_high hurt p_oodc** (9.1 vs 8.8 baseline): raising velocity clamps is counterproductive.

**split_clamp** (single=1.0, tandem=2.0) showed no improvement — the single-foil half regressed on all metrics.

**asym_tan** (single p=2.0, tandem p=3.0) gave mild improvement on p_oodc (8.5) with p_in matching most other variants.

### Memory
~23–24 GB per GPU (3-layer, slice_num=96 — same as R4 baseline).

### Suggested follow-ups
1. **log+highpc at longer horizon**: This run only reached epoch 168; the log transform may need more epochs to converge. Worth running 500+ epochs with `--high_p_clamp --log_pressure --lr 2e-4`.
2. **pclamp=3.0 + log_pressure**: Combine the strongest individual signals (pclamp30 for val/loss, log for p_oodc). Could address both OOD and in-dist regression.
3. **p_in wall analysis**: The consistent p_in regression across all clamp variants suggests the baseline high_p_clamp=2.0 may already be at the optimum. Consider whether p_in can be directly targeted via loss weighting rather than clamp tuning.
4. **Asymmetric clamp with log**: asym_tan + log_pressure could combine the tandem weighting benefit with OOD de-emphasis from log transform.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: R5 review — pclamp30 gives vl=0.2983 (best val/loss) but p_in=13.2. pclamp25 similar. Asymmetric tandem clamp (p_in=13.2, p_ood=8.5) is interesting. None beat baseline p_in=12.7. CLOSE.
